File: Blazor Projects/Pages/Simulation.razor.js
"use strict";

var canvas = document.getElementById("tile-canvas");
var context = canvas.getContext("2d");

var tileSize;
var canvasWidth;
var canvasHeight;

var amountOfRows;
var amountOfColumns;
var dotnetInstance;

registerWindowResizeHandler();
registerClickHandler();

function registerWindowResizeHandler() {
    addEventListener("resize", (e) => { handleWindowResize() });
}

function registerClickHandler() {
    canvas.addEventListener("click", (e) => { handleOnClick(e) });
}

function handleWindowResize() {
    
}

function handleOnClick(e) {
    var tile = canvas.getBoundingClientRect();
    var tileX = e.clientX - tile.left;
    var tileY = e.clientY - tile.top;

    var clickedRow = Math.floor(tileY / tileSize);
    var clickedCol = Math.floor(tileX / tileSize);

    var clickedTile = { X: clickedCol * tileSize, Y: clickedRow * tileSize, Checked: true};

    dotnetInstance.invokeMethodAsync('UpdateTile', JSON.stringify(clickedTile));
}

export function initializeCanvas() {
    canvasWidth = tileSize * amountOfColumns;
    canvasHeight = tileSize * amountOfRows;

    canvas.width = canvasWidth;
    canvas.height = canvasHeight;
}

function clearCanvas() {
    context.clearRect(0, 0, canvasWidth, canvasHeight);
}

export function generateGrid() {
    
    if (tileSize === 0) return;

    initializeCanvas();
    clearCanvas();

    var tiles = [];

    for (var i = 0; i < amountOfRows; i++) {
        for (var j = 0; j < amountOfColumns; j++) {
            var tile = { X: j * tileSize, Y: i * tileSize, Checked:false, LineColor:"gray", FillColor:"transparent", LineWidth: "1px" }
            tiles.push(tile);
        }
    }

    dotnetInstance.invokeMethodAsync('SaveTiles', tiles);
}

export function drawLine(x1, y1, x2, y2, color = "black", lineWidth = "1px") {
    context.beginPath();
    context.moveTo(x1, y1);
    context.lineTo(x2, y2);
    context.lineWidth = lineWidth;
    context.strokeStyle = color;
    context.stroke();
}

export function drawRectangle(x, y, width, height, lineColor = "transparent", lineWidth = "1px", fillColor = "transparent") {
    context.beginPath();
    context.rect(x, y, width, height);
    context.lineWidth = lineWidth;
    context.strokeStyle = lineColor;
    context.stroke();
    context.fillStyle = fillColor;
    context.fill();
}

export function setTileSize(value) {
    tileSize = value;
}

export function setAmountOfColumns(value) {
    amountOfColumns = value;
}

export function setAmountOfRows(value) {
    amountOfRows = value;
}

export function redrawTile(tile) {
    clearRectangle(tile.x, tile.y, tileSize, tileSize);
    drawRectangle(tile.x, tile.y, tileSize, tileSize, tile.lineColor, tile.lineWidth, tile.fillColor);
}

function clearRectangle(x, y, width, height) {
    context.clearRect(x, y, width, height);
}


export function initializeDotNetInstance(instance) {
    dotnetInstance = instance;
}

File: Blazor Projects/Pages/Simulation.razor.cs
using Blazor_Projects.Models;
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using Newtonsoft.Json;

namespace Blazor_Projects.Pages;
public partial class Simulation : BasePage
{
    private IJSObjectReference SimulationScript { get; set; } = null!;

    private int _amountOfColumns = 10;
    private int _maxAmountOfColumns = 100;
    private int _amountOfRows = 10;
    private int _maxAmountOfRows = 100;
    private int _tileSize = 20;
    private List<Tile> _tiles = new();

    protected override async Task OnInitializedAsync()
    {
        await base.OnInitializedAsync();
        await ImportSimulationScript();
        await InitializeDotNetInstanceOnJS();
        await SetGridParametersFromLocalStorage();
        await InitializeCanvas();
        await DrawGrid();
    }

    private async Task ImportSimulationScript()
    {
        SimulationScript = await JS.InvokeAsync<IJSObjectReference>("import", "./Pages/Simulation.razor.js");
        await Task.Delay(1);
    }

    private async Task SetJSVariable(ChangeEventArgs e, string functionName, string variableName)
    {
        var value = Convert.ToInt32(e.Value);
        await SimulationScript.InvokeVoidAsync(functionName, value);

        var gridParameters = await GetGridParametersFromLocalStorage();

        var parameter = gridParameters.GetType().GetProperty(variableName);
        parameter?.SetValue(gridParameters, value);

        await LocalStorage.Set("GridParameters", JsonConvert.SerializeObject(gridParameters));
        await GenerateGrid();
        StateHasChanged();
    }

    private async Task InitializeDotNetInstanceOnJS()
    {
        await SimulationScript.InvokeVoidAsync("initializeDotNetInstance", DotNetObjectReference.Create(this));
    }

    private async Task SetGridParametersFromLocalStorage()
    {
        var gridParameters = await GetGridParametersFromLocalStorage();

        if (gridParameters != null)
        {
            _amountOfColumns = gridParameters.AmountOfColumns;
            _amountOfRows = gridParameters.AmountOfRows;
            _tileSize = gridParameters.TileSize;
        }

        await SimulationScript.InvokeVoidAsync("setTileSize", _tileSize);
        await SimulationScript.InvokeVoidAsync("setAmountOfColumns", _amountOfColumns);
        await SimulationScript.InvokeVoidAsync("setAmountOfRows", _amountOfRows);
    }

    private async Task<GridParameter?> GetGridParametersFromLocalStorage()
    {
        var gridParametersFromLocalStorage = await LocalStorage.Get("GridParameters");
        if (!string.IsNullOrEmpty(gridParametersFromLocalStorage))
        {
            return JsonConvert.DeserializeObject<GridParameter>(gridParametersFromLocalStorage);
        }

        return new GridParameter { AmountOfColumns = _amountOfColumns, AmountOfRows = _amountOfRows, TileSize = _tileSize };
    }

    private async Task DrawGrid()
    {
        var tilesFromLocalStorage = await LocalStorage.Get("Tiles");
        if (!string.IsNullOrEmpty(tilesFromLocalStorage))
        {
            _tiles = JsonConvert.DeserializeObject<List<Tile>>(tilesFromLocalStorage) ?? new List<Tile>();
            if (_tiles.Any())
                await DrawTiles(_tiles);
            else
                await GenerateGrid();
        }
        else
            await GenerateGrid();

    }

    private async Task DrawTiles(List<Tile> tiles)
    {
        foreach (var tile in tiles)
        {
            await SimulationScript.InvokeVoidAsync("drawRectangle", tile.X, tile.Y, _tileSize, _tileSize, tile.LineColor, tile.LineWidth, tile.FillColor);
        }

        StateHasChanged();
    }

    private async Task GenerateGrid()
    {
        await SimulationScript.InvokeVoidAsync("generateGrid");
    }

    private async Task InitializeCanvas()
    {
        await SimulationScript.InvokeVoidAsync("initializeCanvas");
    }
    
    [JSInvokable]
    public async Task SaveTiles(List<Tile> tiles)
    {
        _tiles = tiles;
        await LocalStorage.Set("Tiles", JsonConvert.SerializeObject(_tiles));
        await DrawTiles(_tiles);
    }

    [JSInvokable]
    public async Task UpdateTile(string updatedTileAsJson)
    {
        var updatedTile = JsonConvert.DeserializeObject<Tile>(updatedTileAsJson);
        
        if (updatedTile == null) return;

        var tile = _tiles.FirstOrDefault(x => x.X == updatedTile.X && x.Y == updatedTile.Y);
        if (tile == null) return;

        tile.Checked = updatedTile.Checked;
        tile.FillColor = "gray";

        await SimulationScript.InvokeVoidAsync("redrawTile", tile);
        await LocalStorage.Set("Tiles", JsonConvert.SerializeObject(_tiles));
    }
}
